### PR TITLE
Add Markdown syntax highlighting to the script editor

### LINKED ISSUE #1409: Add support for editing manual/tutorial Markdown files in the script editor
**Describe the project you are working on:**

A dungeon crawler with a dolphin character.

**Describe the problem or limitation you are having in your project:**

I'm having to switch between Godot and a text editor a lot.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**

Rather then having to switch to another program to view and edit text documentation, it would be great to be able to edit and view documentation directly in Godot.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

This proposal is meant for higher-level documentation. Low-level code documentation is a separate concern.

To edit documentation, I imagine this would be almost completely identical to the Script editor (just with different highlighting). I mostly use Markdown for documentation. Here is an example:

image

Optimally there would be a toggle button on the toolbar to switch between editing and viewing. And documentation would be placed alongside Godot documentation tabs.

image

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

Definitely not a few. I have not worked on the editor and am unsure how easy it would be to add.

**Is there a reason why this should be core and not an add-on in the asset library?:**

Optimally everyone should have some kind of documentation to keep their project on track (along with keeping a team in sync). "Failing to plan is planning to fail." Seeing as this is advantageous to everyone, it should be provided to everyone.

I can write several paragraphs on how beneficial documentation is if needed (but please don't make me).

### PR DESCRIPTION
**Pull request by Calinou**
- Follow-up to https://github.com/godotengine/godot/pull/77972 (can be merged independently).

There are some limitations due to how CodeEdit's highlighting works, but it's good enough for a lot of Markdown files out there already. It could be made significantly better if `add_color_region()` had a way to enforce the start to be at the beginning of a line, or with only leading whitespace :slightly_smiling_face: 

- This closes https://github.com/godotengine/godot-proposals/issues/1409.

## Preview

Screenshot_20230616_080330

Screenshot_20230616_080339

Screenshot_20230616_080430

Screenshot_20230616_080452

### CODE REVIEW

**Comment by yoont4** [comment]
Just curious, is there any issue with 4.x compatibility on this PR? Would love to see this in an upcoming release!

**Comment by Calinou** [comment]
> Just curious, is there any issue with 4.x compatibility on this PR? Would love to see this in an upcoming release!

To my knowledge, there's nothing blocking this PR, but it needs a review. Since GDC is ongoing and many core contributors are there, this may take a while.

**Comment by yoont4** [comment]
> To my knowledge, there's nothing blocking this PR, but it needs a review. Since GDC is ongoing and many core contributors are there, this may take a while.

Ah that makes total sense, thanks for the reply!

File: editor/plugins/script_editor_plugin.cpp
Comment (adamscott) [inline editor/plugins/script_editor_plugin.cpp:260]: Is this because the script editor assumes that's GDScript?

Comment (Calinou) [reply editor/plugins/script_editor_plugin.cpp:283]: No, it's because of https://github.com/godotengine/godot/issues/30586 (which still seems to partially exist in 4.x, at least for `_`).

**Review by coppolaemilio — APPROVED** [review]
This is such an easy merge, looking forward to it 💖

**Comment by timothyqiu** [comment]
Some valid markdown might have incorrect highlighting.

image


<details><summary>Text</summary>

```markdown
Autolink is URL between `<` and `>`. See <https://www.example.com/> for details.

Most inline <abbr title="Hypertext Markup Language">HTML</abbr> tags are supported.
```
</details>

We can probably just highlight the `>` at the beginning of a line.

**Comment by Calinou** [comment]
> We can probably just highlight the `>` at the beginning of a line.

`CodeHighligher.add_color_region()` doesn't have a way to make a region effective only if it starts at the beginning of the line, so this would have to be implemented first.

**Comment by Repiteo** [comment]
Thanks!

**Comment by mrfisty** [comment]
Excited for this as well!

**Comment by FrostMiKu** [comment]
Is it possible to export this SyntaxHighlighter to a GDScript class?
I think it would be useful for writing editor programs. Thank you!

**Comment by Calinou** [comment]
> Is it possible to export this SyntaxHighlighter to a GDScript class? I think it would be useful for writing editor programs. Thank you!

I believe this is already exposed to scripting in a way that editor plugins can provide their own highlighters. We may need more documentation for this in the class reference still, or perhaps even a dedicated manual page.

**Pull request merged**
